### ISSUE TITLE
Help it is just a black box.

### ISSUE THREAD

**Issue by sbottingota** [firefish111/os#1]
I am too stupid to get stuff to work on this os. The only way i have been able to run code on it so far is to copy it into one of the kernel files, but that's just f---ing stupid there should be a better way. Please help.

![image](https://github.com/user-attachments/assets/ede1276b-8e82-413e-aaca-195e2a60b264)

<details>
<summary>Click here to show/hide console output</summary>

```
saku@porrigecourt:~/MiscProjects/OSTest/far-os$ make bochs
bochs -q
00000000000i[      ] LTDL_LIBRARY_PATH not set. using compile time default '/usr/lib/x86_64-linux-gnu/bochs/plugins'
========================================================================
                        Bochs x86 Emulator 2.7
              Built from SVN snapshot on August  1, 2021
                Timestamp: Sun Aug  1 10:07:00 CEST 2021
========================================================================
00000000000i[      ] BXSHARE not set. using compile time default '/usr/share/bochs'
00000000000i[      ] lt_dlhandle is 0x5c16e17f3700
00000000000i[PLUGIN] loaded plugin libbx_extfpuirq.so
00000000000i[      ] lt_dlhandle is 0x5c16e17f3890
00000000000i[PLUGIN] loaded plugin libbx_unmapped.so
00000000000i[      ] lt_dlhandle is 0x5c16e17a4a90
00000000000i[PLUGIN] loaded plugin libbx_parallel.so
00000000000i[      ] lt_dlhandle is 0x5c16e17a66b0
00000000000i[PLUGIN] loaded plugin libbx_biosdev.so
00000000000i[      ] lt_dlhandle is 0x5c16e17a70c0
00000000000i[PLUGIN] loaded plugin libbx_serial.so
00000000000i[      ] lt_dlhandle is 0x5c16e17bc830
00000000000i[PLUGIN] loaded plugin libbx_iodebug.so
00000000000i[      ] lt_dlhandle is 0x5c16e17bd140
00000000000i[PLUGIN] loaded plugin libbx_speaker.so
00000000000i[      ] lt_dlhandle is 0x5c16e17bf1e0
00000000000i[PLUGIN] loaded plugin libbx_gameport.so
00000000000i[      ] reading configuration from .bochsrc
00000000000i[      ] lt_dlhandle is 0x5c16e17bfcb0
00000000000i[PLUGIN] loaded plugin libbx_textconfig.so
00000000000i[      ] lt_dlhandle is 0x5c16e17c0480
00000000000i[PLUGIN] loaded plugin libbx_x_gui.so
00000000000i[      ] installing x module as the Bochs GUI
00000000000i[      ] Bochs x86 Emulator 2.7
00000000000i[      ]   Built from SVN snapshot on August  1, 2021
00000000000i[      ]   Timestamp: Sun Aug  1 10:07:00 CEST 2021
00000000000i[      ] System configuration
00000000000i[      ]   processors: 1 (cores=1, HT threads=1)
00000000000i[      ]   A20 line support: yes
00000000000i[      ] IPS is set to 4000000
00000000000i[      ] CPU configuration
00000000000i[      ]   SMP support: yes, quantum=16
00000000000i[      ]   Using pre-defined CPU configuration: p4_willamette
00000000000i[      ] Optimization configuration
00000000000i[      ]   RepeatSpeedups support: yes
00000000000i[      ]   Fast function calls: yes
00000000000i[      ]   Handlers Chaining speedups: no
00000000000i[      ] Devices configuration
00000000000i[      ]   PCI support: i440FX i430FX i440BX
00000000000i[      ]   Network devices support: NE2000 E1000
00000000000i[      ]   Sound support: SB16 ES1370
00000000000i[      ]   USB support: UHCI OHCI EHCI xHCI
00000000000i[      ]   VGA extension support: vbe cirrus voodoo
00000000000i[IMG   ] Disk image modules
00000000000i[IMG   ]   flat concat sparse dll growing undoable volatile vmware3
00000000000i[IMG   ]   vvfat vmware4 vpc vbox
00000000000i[NETCTL] Networking modules
00000000000i[NETCTL]   null vde linux tuntap slirp vnet socket
00000000000i[SNDCTL] Sound drivers
00000000000i[SNDCTL]   dummy alsa oss file
00000000000i[USBCTL] Pluggable USB devices
00000000000i[USBCTL]   printer disk cdrom hub floppy mouse tablet keypad
00000000000i[USBCTL]   keyboard
00000000000i[MEM0  ] allocated memory at 0x79102f000010. after alignment, vector=0x79102f001000
00000000000i[MEM0  ] 32.00MB
00000000000i[MEM0  ] mem block size = 0x00020000, blocks=256
00000000000i[MEM0  ] rom at 0xfffe0000/131072 ('/usr/share/bochs/BIOS-bochs-latest')
00000000000i[DEV   ] will paste characters every 100 iodev timer ticks
00000000000i[      ] lt_dlhandle is 0x5c16e1e22270
00000000000i[PLUGIN] loaded plugin libbx_soundalsa.so
00000000000i[WAVOUT] ALSA: opened default PCM output device
00000000000i[WAVOUT] changed sample rate to 44099
00000000000i[      ] lt_dlhandle is 0x5c16e1eb3090
00000000000i[PLUGIN] loaded plugin libbx_pci.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb33c0
00000000000i[PLUGIN] loaded plugin libbx_pci2isa.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb31f0
00000000000i[PLUGIN] loaded plugin libbx_usb_uhci.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb7590
00000000000i[PLUGIN] loaded plugin libbx_acpi.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb80c0
00000000000i[PLUGIN] loaded plugin libbx_hpet.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb8f70
00000000000i[PLUGIN] loaded plugin libbx_cmos.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb97b0
00000000000i[PLUGIN] loaded plugin libbx_dma.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eb9ff0
00000000000i[PLUGIN] loaded plugin libbx_pic.so
00000000000i[      ] lt_dlhandle is 0x5c16e1eba850
00000000000i[PLUGIN] loaded plugin libbx_pit.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebb240
00000000000i[PLUGIN] loaded plugin libbx_vga.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebbaf0
00000000000i[PLUGIN] loaded plugin libbx_floppy.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebc6e0
00000000000i[PLUGIN] loaded plugin libbx_ioapic.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebd040
00000000000i[PLUGIN] loaded plugin libbx_keyboard.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebd970
00000000000i[PLUGIN] loaded plugin libbx_harddrv.so
00000000000i[      ] lt_dlhandle is 0x5c16e1ebfa90
00000000000i[PLUGIN] loaded plugin libbx_pci_ide.so
00000000000i[PLUGIN] init_dev of 'pci' plugin device by virtual method
00000000000i[DEV   ] i440FX PMC present at device 0, function 0
00000000000i[PLUGIN] init_dev of 'pci2isa' plugin device by virtual method
00000000000i[DEV   ] PIIX3 PCI-to-ISA bridge present at device 1, function 0
00000000000i[PLUGIN] init_dev of 'cmos' plugin device by virtual method
00000000000i[CMOS  ] Using local time for initial clock
00000000000i[CMOS  ] Setting initial clock to: Fri Sep 13 17:55:28 2024 (time0=1726246528)
00000000000i[PLUGIN] init_dev of 'dma' plugin device by virtual method
00000000000i[DMA   ] channel 4 used by cascade
00000000000i[PLUGIN] init_dev of 'pic' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'pit' plugin device by virtual method
00000000000i[PIT   ] PIT using realtime synchronisation method
00000000000i[PLUGIN] init_dev of 'vga' plugin device by virtual method
00000000000i[MEM0  ] Register memory access handlers: 0x0000000a0000 - 0x0000000bffff
00000000000i[VGA   ] interval=200000, mode=realtime
00000000000i[VGA   ] VSYNC using realtime mode
00000000000i[MEM0  ] Register memory access handlers: 0x0000e0000000 - 0x0000e0ffffff
00000000000i[BXVGA ] VBE Bochs Display Extension Enabled
00000000000i[XGUI  ] disabled host keyboard repeat
00000000000i[XGUI  ] test_alloc_colors: 16 colors available out of 16 colors tried
00000000000i[XGUI  ] font 8 wide x 16 high, display depth = 24
00000000000i[KEYMAP] Loading keymap from '/usr/share/bochs/keymaps/x11-pc-uk.map'
00000000000i[KEYMAP] Loaded 172 symbols
00000000000i[XGUI  ] maximum host resolution: x=1920 y=1030
00000000000i[MEM0  ] rom at 0xc0000/38400 ('/usr/share/bochs/VGABIOS-lgpl-latest')
00000000000i[PLUGIN] init_dev of 'floppy' plugin device by virtual method
00000000000i[DMA   ] channel 2 used by Floppy Drive
00000000000i[FLOPPY] Using boot sequence disk, none, none
00000000000i[FLOPPY] Floppy boot signature check is enabled
00000000000i[PLUGIN] init_dev of 'acpi' plugin device by virtual method
00000000000i[DEV   ] ACPI Controller present at device 1, function 3
00000000000i[PLUGIN] init_dev of 'hpet' plugin device by virtual method
00000000000i[HPET  ] initializing HPET
00000000000i[MEM0  ] Register memory access handlers: 0x0000fed00000 - 0x0000fed003ff
00000000000i[PLUGIN] init_dev of 'ioapic' plugin device by virtual method
00000000000i[IOAPIC] initializing I/O APIC
00000000000i[MEM0  ] Register memory access handlers: 0x0000fec00000 - 0x0000fec00fff
00000000000i[IOAPIC] IOAPIC enabled (base address = 0xfec00000)
00000000000i[PLUGIN] init_dev of 'keyboard' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'harddrv' plugin device by virtual method
00000000000i[HD    ] HD on ata0-0: 'os.img', 'flat' mode
00000000000e[IMG   ] image locked: 'os.img'
00000000000p[HD    ] >>PANIC<< ata0-0: could not open hard drive image file 'os.img'
00000000000i[PLUGIN] init_dev of 'pci_ide' plugin device by virtual method
00000000000i[DEV   ] PIIX3 PCI IDE controller present at device 1, function 1
00000000000i[PLUGIN] init_dev of 'extfpuirq' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'unmapped' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'parallel' plugin device by virtual method
00000000000i[PAR   ] parallel port 1 at 0x0378 irq 7
00000000000i[PLUGIN] init_dev of 'biosdev' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'serial' plugin device by virtual method
00000000000i[SER   ] com1 at 0x03f8 irq 4 (mode: null)
00000000000i[PLUGIN] init_dev of 'iodebug' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'speaker' plugin device by virtual method
00000000000i[PCSPK ] Using lowlevel sound support for output
00000000000i[PLUGIN] init_dev of 'gameport' plugin device by virtual method
00000000000i[PLUGIN] init_dev of 'usb_uhci' plugin device by virtual method
00000000000i[DEV   ] USB UHCI present at device 1, function 2
00000000000i[UHCI  ] USB UHCI initialized
00000000000i[PLUGIN] register state of 'pci' plugin device by virtual method
00000000000i[PLUGIN] register state of 'pci2isa' plugin device by virtual method
00000000000i[PLUGIN] register state of 'cmos' plugin device by virtual method
00000000000i[PLUGIN] register state of 'dma' plugin device by virtual method
00000000000i[PLUGIN] register state of 'pic' plugin device by virtual method
00000000000i[PLUGIN] register state of 'pit' plugin device by virtual method
00000000000i[PLUGIN] register state of 'vga' plugin device by virtual method
00000000000i[PLUGIN] register state of 'floppy' plugin device by virtual method
00000000000i[PLUGIN] register state of 'extfpuirq' plugin device by virtual method
00000000000i[PLUGIN] register state of 'unmapped' plugin device by virtual method
00000000000i[PLUGIN] register state of 'parallel' plugin device by virtual method
00000000000i[PLUGIN] register state of 'biosdev' plugin device by virtual method
00000000000i[PLUGIN] register state of 'serial' plugin device by virtual method
00000000000i[PLUGIN] register state of 'iodebug' plugin device by virtual method
00000000000i[PLUGIN] register state of 'speaker' plugin device by virtual method
00000000000i[PLUGIN] register state of 'gameport' plugin device by virtual method
00000000000i[PLUGIN] register state of 'usb_uhci' plugin device by virtual method
00000000000i[PLUGIN] register state of 'acpi' plugin device by virtual method
00000000000i[PLUGIN] register state of 'hpet' plugin device by virtual method
00000000000i[PLUGIN] register state of 'ioapic' plugin device by virtual method
00000000000i[PLUGIN] register state of 'keyboard' plugin device by virtual method
00000000000i[PLUGIN] register state of 'harddrv' plugin device by virtual method
00000000000i[PLUGIN] register state of 'pci_ide' plugin device by virtual method
00000000000i[SYS   ] bx_pc_system_c::Reset(HARDWARE) called
00000000000i[CPU0  ] cpu hardware reset
00000000000i[APIC0 ] allocate APIC id=0 (MMIO enabled) to 0x0000fee00000
00000000000i[CPU0  ] CPU[0] is the bootstrap processor
00000000000i[CPU0  ] CPUID[0x00000000]: 00000002 756e6547 6c65746e 49656e69
00000000000i[CPU0  ] CPUID[0x00000001]: 00000f12 00010800 00000000 1febfbff
00000000000i[CPU0  ] CPUID[0x00000002]: 665b5001 00000000 00000000 007a7040
00000000000i[CPU0  ] CPUID[0x80000000]: 80000004 00000000 00000000 00000000
00000000000i[CPU0  ] CPUID[0x80000001]: 00000000 00000000 00000000 00000000
00000000000i[CPU0  ] CPUID[0x80000002]: 20202020 20202020 20202020 6e492020
00000000000i[CPU0  ] CPUID[0x80000003]: 286c6574 50202952 69746e65 52286d75
00000000000i[CPU0  ] CPUID[0x80000004]: 20342029 20555043 30382e31 007a4847
00000000000i[CPU0  ] CPU Features supported:
00000000000i[CPU0  ] 		x87
00000000000i[CPU0  ] 		486ni
00000000000i[CPU0  ] 		pentium_ni
00000000000i[CPU0  ] 		p6ni
00000000000i[CPU0  ] 		mmx
00000000000i[CPU0  ] 		debugext
00000000000i[CPU0  ] 		vme
00000000000i[CPU0  ] 		pse
00000000000i[CPU0  ] 		pae
00000000000i[CPU0  ] 		pge
00000000000i[CPU0  ] 		pse36
00000000000i[CPU0  ] 		mtrr
00000000000i[CPU0  ] 		pat
00000000000i[CPU0  ] 		sysenter_sysexit
00000000000i[CPU0  ] 		clflush
00000000000i[CPU0  ] 		sse
00000000000i[CPU0  ] 		sse2
00000000000i[CPU0  ] 		xapic
00000000000i[PLUGIN] reset of 'pci' plugin device by virtual method
00000000000i[PLUGIN] reset of 'pci2isa' plugin device by virtual method
00000000000i[PLUGIN] reset of 'cmos' plugin device by virtual method
00000000000i[PLUGIN] reset of 'dma' plugin device by virtual method
00000000000i[PLUGIN] reset of 'pic' plugin device by virtual method
00000000000i[PLUGIN] reset of 'pit' plugin device by virtual method
00000000000i[PLUGIN] reset of 'vga' plugin device by virtual method
00000000000i[PLUGIN] reset of 'floppy' plugin device by virtual method
00000000000i[PLUGIN] reset of 'acpi' plugin device by virtual method
00000000000i[PLUGIN] reset of 'hpet' plugin device by virtual method
00000000000i[PLUGIN] reset of 'ioapic' plugin device by virtual method
00000000000i[PLUGIN] reset of 'keyboard' plugin device by virtual method
00000000000i[PLUGIN] reset of 'harddrv' plugin device by virtual method
00000000000i[PLUGIN] reset of 'pci_ide' plugin device by virtual method
00000000000i[PLUGIN] reset of 'extfpuirq' plugin device by virtual method
00000000000i[PLUGIN] reset of 'unmapped' plugin device by virtual method
00000000000i[PLUGIN] reset of 'parallel' plugin device by virtual method
00000000000i[PLUGIN] reset of 'biosdev' plugin device by virtual method
00000000000i[PLUGIN] reset of 'serial' plugin device by virtual method
00000000000i[PLUGIN] reset of 'iodebug' plugin device by virtual method
00000000000i[PLUGIN] reset of 'speaker' plugin device by virtual method
00000000000i[PLUGIN] reset of 'gameport' plugin device by virtual method
00000000000i[PLUGIN] reset of 'usb_uhci' plugin device by virtual method
00000000000i[      ] set SIGINT handler to bx_debug_ctrlc_handler
Next at t=0
(0) [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
<bochs:1> exit
00000000000i[      ] dbg: Quit
00000000000i[CPU0  ] CPU is in real mode (active)
00000000000i[CPU0  ] CS.mode = 16 bit
00000000000i[CPU0  ] SS.mode = 16 bit
00000000000i[CPU0  ] EFER   = 0x00000000
00000000000i[CPU0  ] | EAX=00000000  EBX=00000000  ECX=00000000  EDX=00000000
00000000000i[CPU0  ] | ESP=00000000  EBP=00000000  ESI=00000000  EDI=00000000
00000000000i[CPU0  ] | IOPL=0 id vip vif ac vm rf nt of df if tf sf zf af pf cf
00000000000i[CPU0  ] | SEG sltr(index|ti|rpl)     base    limit G D
00000000000i[CPU0  ] |  CS:f000( 1e00| 0|  0) ffff0000 0000ffff 0 0
00000000000i[CPU0  ] |  DS:0000( 0000| 0|  0) 00000000 0000ffff 0 0
00000000000i[CPU0  ] |  SS:0000( 0000| 0|  0) 00000000 0000ffff 0 0
00000000000i[CPU0  ] |  ES:0000( 0000| 0|  0) 00000000 0000ffff 0 0
00000000000i[CPU0  ] |  FS:0000( 0000| 0|  0) 00000000 0000ffff 0 0
00000000000i[CPU0  ] |  GS:0000( 0000| 0|  0) 00000000 0000ffff 0 0
00000000000i[CPU0  ] | EIP=0000fff0 (0000fff0)
00000000000i[CPU0  ] | CR0=0x60000010 CR2=0x00000000
00000000000i[CPU0  ] | CR3=0x00000000 CR4=0x00000000
(0).[0] [0x0000fffffff0] f000:fff0 (unk. ctxt): jmpf 0xf000:e05b          ; ea5be000f0
00000000000i[CMOS  ] Last time is 1726246528 (Fri Sep 13 17:55:28 2024)
00000000000i[XGUI  ] Exit
00000000000i[SIM   ] quit_sim called with exit code 0
```
</details>

**Comment by firefish111** [comment]
Yeah, bochs is weird like that, it starts in debug mode so you'd need to type `continue` (`c` for short) to start execution
I would say just use QEMU, it's much faster anyway

**Comment by sbottingota** [comment]
Cool works.

At 400WPM without sleeping, it should only take me until halfway through trinity to make something that is remotely similar in ability :)

btw if i were you i would add to the readme that:
- qic does not come packaged with the thing, so you will have to clone it separately (at least I had to).
  <details>
  <summary></summary>

  ```
    saku@porrigecourt:~/MiscProjects/OSTest$ git clone git@github.com:far-os/os.git
  Cloning into 'os'...
  remote: Enumerating objects: 724, done.
  remote: Counting objects: 100% (451/451), done.
  remote: Compressing objects: 100% (312/312), done.
  remote: Total 724 (delta 317), reused 260 (delta 129), pack-reused 273 (from 1)
  Receiving objects: 100% (724/724), 195.13 KiB | 880.00 KiB/s, done.
  Resolving deltas: 100% (491/491), done.
  saku@porrigecourt:~/MiscProjects/OSTest$ ls os/util/qic/
  saku@porrigecourt:~/MiscProjects/OSTest$ 
  ```
  </details>

- Apparently debian based operating systems like to install bochs in little bits, so you will have to install `bochs-x` separately using apt (or some other package manager i guess)
- You have changed your recomendation from bochs to the other one.

**Comment by firefish111** [comment]
you shouldnt need to clone qic seperately, just clone with `--recursive`

the only reason I had recommended bochs in the past was because there was a race condition to do with interrupt setup that is now fixed (see commit `61da6d9`'s changes to kernel.c) that seemed to only affect qemu

and yes i will add that all to readme next time i make a change to the code at all (am lazy)
